gdsfactory/simulation/gmeep/get_sparameters.py

```diff
@@ -374,9 +307,16 @@ def sparameter_calculation(
     # for port_index in Sparams_indices:
 
     if lazy_parallelism:
+        import multiprocessing
+
         from mpi4py import MPI
 
-        n = mp.divide_parallel_processes(len(Sparams_indices))
+        cores = min([len(Sparams_indices), multiprocessing.cpu_count()])
+        # mp.count_processors = lambda x: cores
+        # FIXME RuntimeError: meep: numgroups > count_processors
+        # Using MPI version 3.1, 1 processes
```

Comment:
Hi @smartalecH  and @simbilod,

how can i configure the mpi meep to use more than 1 process?

Reply:
As long as you have the mpi build of meep installed on your computer, you can run your simulation on multiple processes using:
```
mpirun python3 my_script.py
```
This will launch as many processes as you have cores on your system. Alternatively, if you only want to use 2 of your cores (rather than your whole system) you could explicitly pass the number of cores to use:
```
mpirun -np 2 python3 my_script.py
```
There's an important subtlety to MPI. MPI will literally launch the same script on each process. Since you are planning to do a meep simulation within gdsFactory, your gdsFactory routine will run `N` times (in parallel). This means you will use `N` times as much memory -- *except* when you get to the meep code. Meep was setup, of course, to leverage MPI under the hood. So it will distribute the simulation accordingly and accelerate (because now each process is stepping a *different* section/chunk of the computation domain).

People doing large-scale optimization (including me) run into issues caused by this all the time. Our python optimization code is replicated on each process. If the algorithm uses a lot of storage, then this approach doesn't scale well.

In your case, I don't think it'll be an issue. Just something to be aware of.

(note that Meep also supports thread-level parallelism, but it's significantly slower than MPI right now -- there's a lot of work to be done there).

Reply:
To add to Alec`s excellent breakdown, the package `mpi4py` has tools to make different (sets of) processors see different forms of `myscript.py` when run via `mpirun python myscript.py`. 

This is what is done in the `get_sparametersNxN` script toward the end to merge the different results from the separate runs generated by `divide_parallel_processes()`:

```
# Setup
from mpi4py import MPI
n = mp.divide_parallel_processes(Nprocess)
comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

# Perform conditional MEEP simulation + analyze results
results[n] = function(n) # Here, results is a Dict whose keys depend on n

# Conditional operations depending on the process
if rank == 0:
    for i in range(1, size, 1):
        data = comm.recv(source=i, tag=11)
        results.update(data)

        return results
else:
    comm.send(results, dest=0, tag=11)
```

If the script above is run via `mpirun -np Nprocessors python script.py`, all N=`Nprocessors` processors will execute a copy of this entire script. When running the `divide_parallel_processes(Nprocesses)` line, the set of Nprocessors processors will be subdivided into M=`Nprocesses` groups, each using Nprocessors/Nprocesses. Furthermore  `n` will be assigned a different value depending on the group. This makes it so when doing a MEEP omputation, `results` can be made to depend on `n` -- in my case, I made that set the source location.

Now the tricky part comes when you want to operate on the aggregate results. Each subprocess has its own set of variables in memory, including `results`, and so returning `results` will only return the `results` variable from one of the subprocesses. `mpi4py` can add more control flow that depends on the process: here, the script will stop, and process #0 will wait for communication from the other processes to generate one big dictionary that combines the results across all simulations.

Reply:
I'll just add that meep wraps quite a few of the above-mentioned tools. `mpi4py` has quite a bit of boilerplate, so it's nice to avoid when possible:

- [`divide_parallel_processes()`](https://github.com/NanoComp/meep/blob/efe3b4b1fff538cf1b1ad2e34bb47c7a3ddfb28a/src/mympi.cpp#L820-L831)
- [`get_num_groups()`](https://github.com/NanoComp/meep/blob/efe3b4b1fff538cf1b1ad2e34bb47c7a3ddfb28a/python/simulation.py#L5390-L5395)
- [`get_group_masters()`](https://github.com/NanoComp/meep/blob/efe3b4b1fff538cf1b1ad2e34bb47c7a3ddfb28a/python/simulation.py#L5397-L5418)
- [`merge_subgroup_data()`](https://github.com/NanoComp/meep/blob/efe3b4b1fff538cf1b1ad2e34bb47c7a3ddfb28a/python/simulation.py#L5420-L5460)